core/bundle/changes/changes.go

```diff
@@ -1225,29 +1225,110 @@ func (cs *changeset) dependents() map[string][]string {
 }
 
 // sorted returns the changes sorted by requirements, required first.
-func (cs *changeset) sorted() []Change {
-	done := set.NewStrings()
-	var sorted []Change
-	changes := cs.changes[:]
-mainloop:
-	for len(changes) != 0 {
-		// Note that all valid bundles have at least two changes
-		// (add one charm and deploy one application).
-		change := changes[0]
-		changes = changes[1:]
-		for _, r := range change.Requires() {
-			if !done.Contains(r) {
-
-				// This change requires a change which is not yet listed.
-				// Push this change at the end of the list and retry later.
-				changes = append(changes, change)
-				continue mainloop
+func (cs *changeset) sorted() ([]Change, error) {
+	// Exit it early if there is nothing to sort.
+	if len(cs.changes) == 0 {
+		return []Change{}, nil
+	}
+
+	// Create a map to sort the data.
+	data := make(map[string][]string, len(cs.changes))
+	dataOrder := make([]string, len(cs.changes))
+	for i, c := range cs.changes {
+		// Sorting the requirements can be removed if the place in
+		// handlers.go where the order of requirements is not
+		// idempotent can be found and fixed.
+		sort.Sort(sort.StringSlice(c.Requires()))
+		data[c.Id()] = c.Requires()
+		dataOrder[i] = c.Id()
+	}
+	sortedChangeIDs, err := toposort_flatten(dataOrder, data)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// convert the sorted change IDs to a sorted slice
+	// of changes
+	var sortedChanges []Change
+	for _, changeID := range sortedChangeIDs {
+		for _, change := range cs.changes {
+			if changeID == change.Id() {
+				sortedChanges = append(sortedChanges, change)
+				break
+			}
+		}
+	}
+	return sortedChanges, nil
+}
+
+// toposort_flatten performs a topographical flattened sort on the provided
+// data.  dataOrder is a slice of the originally ordered changes. data is a
+// map of change to requirements.  To be idempotent, requirements must be in
+// the same order each time.  Use along with data to ensure that this method
+// is deterministic.
+//
+// Adapted from https://tylercipriani.com/blog/2017/09/13/topographical-sorting-in-golang/
+func toposort_flatten(dataOrder []string, data map[string][]string) ([]string, error) {
```

Comment:
the function name should probably be topoSortFlatten

Reply:
also would be nice to add a test for this function directly.

Reply:
Also would be good to attribute with some more information: CC BY-SA 4.0 "If supplied, you must provide the name of the creator and attribution parties, a copyright notice, a license notice, a disclaimer notice, and a link to the material. CC licenses prior to Version 4.0 also require you to provide the title of the material if supplied, and may have other slight differences."

Reply:
tests are in changessort_test.go